src/python/pants/engine/build_files.py

```diff
@@ -201,90 +203,93 @@ def _hydrate(item_type, spec_path, **kwargs):
   return item
 
 
-@rule(BuildFileAddresses, [Select(AddressMapper), Select(Specs)])
-def addresses_from_address_families(address_mapper, specs):
-  """Given an AddressMapper and list of Specs, return matching BuildFileAddresses.
-
-  Raises a AddressLookupError if:
+class MappedSpecs(datatype([
+    ('address_families', tuple),
+    ('specs', Specs),
+])):
+  """Wraps up logic to resolve specs into individual target addresses, or error out."""
+
+  def __new__(cls, address_families, specs):
+    return super(MappedSpecs, cls).__new__(cls, tuple(address_families), specs)
+
+  @memoized_property
```

Comment:
I don't think there is really any advantage to memoizing these on the class, but the downside is that all of these memoized properties will survive on the object between runs, because the `MappedSpecs` object is itself memoized.

If you were to undo the @rule split, it would be a non-issue, because the `MappedSpecs` object would be a local-only construct. But as it stands, you're holding on to a lot of stuff that could/should be ephemeral.

Reply:
I'm not too attached to the `@rule` split (lol). I'll try merging them back, I don't think it's necessary for this to work.

Reply:
Btw, this was a very helpful explanation re: memoization. Thanks.

Reply:
The rules have been merged back together into one, maintaining the `MappedSpecs` object, but not having it anywhere inside the rule graph (and therefore memoized/etc).